File: src/sp/it/pl/util/Hierarchical.java
package sp.it.pl.util;

import java.util.List;
import java.util.stream.Stream;
import sp.it.pl.util.functional.Util;
import static sp.it.pl.util.functional.UtilKt.seqRec;

/**
 * Object in a tree hierarchy, having an optional parent and any number of children. All objects in the hierarchy
 * inherit from <i>hierarchy</i> type H.
 *
 * @param <H> hierarchy type
 */
public interface Hierarchical<H extends Hierarchical<H>> {

	H getHParent();

	List<H> getHChildren();

	@SuppressWarnings("unchecked")
	default Stream<H> getHChildrenR() {
		return Util.stream(seqRec((H) this, e -> e.getHChildren()).iterator());
	}

	default boolean isHLeaf() {
		return getHChildren().isEmpty();
	}

	default Stream<H> getHLeafChildrenR() {
		return getHChildrenR().filter(Hierarchical::isHLeaf);
	}

	default boolean isHRoot() {
		return getHParent()==null;
	}

	@SuppressWarnings("unchecked")
	default H getHRoot() {
		H i = (H) this, p;

		// Depending on implementation, this may perform sub-optimally as it calls getHParent twice
		// Hence, we optimize it to the below
		// while (!i.isHRoot())
		//	  i = i.getHParent();

		while ((p = i.getHParent())!=null)
			i = p;

		return i;
	}

	@SuppressWarnings("unchecked")
	default boolean isHChildOf(H h) {
		if (this==h) return false;
		H i = (H) this;

		while ((i = i.getHParent())!=null)
			if (i==h) return true;
		return false;
	}

	@SuppressWarnings("unchecked")
	default boolean isHParentOf(H h) {
		return h!=null && h.isHChildOf((H) this);
	}

}

File: src/util/main/sp/it/util/Hierarchical.kt
package sp.it.util

import sp.it.util.functional.recurse
import sp.it.util.functional.traverse

/**
 * Object in a tree hierarchy, having an optional parent and any number of children.
 * All objects in the hierarchy inherit from *hierarchy* type H.
 *
 * @param <H> hierarchy type
 */
@Suppress("UNCHECKED_CAST")
interface Hierarchical<H: Hierarchical<H>> {
   /** Hierarchy parent of this hierarchical item */
   val hParent: H?
   /** Hierarchy children of this hierarchical item */
   val hChildren: List<H>
   /** Hierarchy traversal sequence of direct children of this hierarchical item */
   val hChildrenTraverse: Sequence<H>
      get() = hChildren.asSequence()
   /** Hierarchy traversal recursive sequence of all children of this hierarchical item including this as first element */
   val hChildrenRecurse: Sequence<H>
      get() = (this as H).recurse { it.hChildren }
   /** Hierarchy traversal recursive sequence of all leaf children of this hierarchical item including this as first element */
   val hChildrenRecurseLeafs: Sequence<H>
      get() = hChildrenRecurse.filter { it.isHLeaf }
   /** Whether this hierarchical item is oot of the hierarchy */
   val isHRoot: Boolean
      get() = hParent==null
   /** Whether this hierarchical item is leaf - has no children */
   val isHLeaf: Boolean
      get() = hChildren.isEmpty()
   /** Hierarchy root of this hierarchical item */
   val hRoot: H?
      get() = (this as H).traverse { it.hParent }.last()

   fun isHChildOf(h: H): Boolean = if (this===h) false else (this as H).traverse { it.hParent }.any { it===h }

   fun isHParentOf(h: H?): Boolean = h!=null && h.isHChildOf(this as H)

}